### ISSUE TITLE
Dada2 pipeline

### ISSUE THREAD

**Issue by annarizzibramucci** [benjjneb/dada2#944]
I am new to Dada2 and I am having issues because my filtering step is removing sequences before the derep steps. I have looked at your trouble shooting for this and found several suggestions but I cannot get any of them to work. I have my filtN files going into a filtN folder and then I was attempting to make new sample lists so that I just call up those files, then so on inside of each subsequent folder.

example-------------

fnFs <- sort(list.files(data.fp, pattern="L001_R1.fastq", full.names = TRUE));
fnRs <- sort(list.files(data.fp, pattern="L001_R2.fastq", full.names = TRUE));
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1);

data.fp <- ("..../fAHG6G/")
fnFs.filtN <- file.path(data.fp, "filtN", basename(fnFs)); # Put N-filterd files in filtN/ subdirectory
fnRs.filtN <- file.path(data.fp, "filtN", basename(fnRs));

filterAndTrim(fnFs, fnFs.filtN, fnRs, fnRs.filtN, maxN = 0, multithread = TRUE, compress=F);

data.fp <- ("..../filtN/cutadapt")

# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
cutFs <- sort(list.files(data.fp, pattern="_R1.fastq", full.names = TRUE));
cutRs <- sort(list.files(data.fp, pattern="_R2.fastq", full.names = TRUE));
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(cutFs), "_"), `[`, 1);

So basically I am trying to get nested folders AHG6G>filtN>cutadapt>trimmed>AHG6Gdada2

Then when I start a new batch of code within in each folder I am just trying to make a new list of files and then move on with the tutorial using the new lists

thanks in advance.
cheers
Anna

**Comment by annarizzibramucci** [comment]
The other thing that I really want to know how to fix is an issue with cut adapt, which is that when I remove my primers I am getting rid of most of the primers but about 11 remaining primers are present in FWD and REV sequences, and I am wondering your advice on how to deal with this issue? should I just run cut adapt twice or should I just remove those sequences at the end after merging?
thanks again!

**Comment by benjjneb** [comment]
The standard idiom I suggest to cut down to just the files that were output is to use the `file.exists` command. So, if you want to get a list of just the `filtN` files that were successfully output:

```
filtN <- filtN[file.exists(filtN)]
```

That will remove any entries in your original list of samples that weren't output at that step because zero reads passed the filter.

**Comment by benjjneb** [comment]
> but about 11 remaining primers are present in FWD and REV sequences

Assuming this is far, far lower than the number of reads in your samples, this is probably safe to ignore. A small number of such unremoved primer matches might show up due to PCR artifacts that put the primer sequence in an unexpected place in the read, but they will (mostly) be removed by the pipeline so not a concern.

**Comment by annarizzibramucci** [comment]
My code is still falling over here...

data.fp <- ("/Users/annarizzibramucci/Downloads/australian-microbiome-2/AHG6G/filtN/cutadapt")
cutFs <- cutadapt[file.exists(cutadapt)]
cutRs <- cutadapt[file.exists(cutadapt)]

trimFs <- file.path(data.fp, "trimmed", paste0(basename(fnFs), "_R1_trim.fastq"))
trimRs <- file.path(data.fp, "trimmed", paste0(basename(fnFs), "_R2_trim.fastq"))

out <- filterAndTrim(cutFs, trimFs, cutRs, trimRs, truncLen=c(250,228),
              maxN=0, maxEE=c(6,10), truncQ=2, rm.phix=TRUE,
              compress=FALSE, multithread=TRUE, minLen = 50, matchIDs=TRUE); # On Windows set multithread=FALSE
head(out);

**Comment by benjjneb** [comment]
How is it failing? What is the error message?

My guess is that the trimFs vector of filenames you defined don't match the cutFs because you built them from the fnFs which include the files that were later removed in filtering. If you build them from the cutFs you are probbaly fine, i.e.

```
trimFs <- file.path(data.fp, "trimmed", paste0(basename(**cutFs**), "_R1_trim.fastq"))
```

**Comment by annarizzibramucci** [comment]
The error message is:
Error in filterAndTrim(cutFs, trimFs, cutRs, trimRs, truncLen = c(250,  : 
  Every input file must have a corresponding output file.

**Comment by annarizzibramucci** [comment]
So actually I tried exactly what you suggested previously and got the following error:
Error in (function (fn, fout, maxN = c(0, 0), truncQ = c(2, 2), truncLen = c(0,  : 
  The output and input file names must be different.

In that attempt my code was:
data.fp <- ("/Users/annarizzibramucci/Downloads/australian-microbiome-2/AHG6G/filtN/cutadapt")
cutFs <- cutadapt[file.exists(cutadapt)]
cutRs <- cutadapt[file.exists(cutadapt)]

trimFs <- file.path(data.fp, "trimmed", paste0(basename(cutFs), "_R1_trim.fastq"))
trimRs <- file.path(data.fp, "trimmed", paste0(basename(cutRs), "_R2_trim.fastq"))

out <- filterAndTrim(cutFs, trimFs, cutRs, trimRs, truncLen=c(250,228),
              maxN=0, maxEE=c(6,10), truncQ=2, rm.phix=TRUE,
              compress=FALSE, multithread=TRUE, minLen = 50, matchIDs=TRUE); # On Windows set multithread=FALSE
head(out);

**Comment by benjjneb** [comment]
What is the output of

```
head(cutFs)
head(trimFs)
```

**Comment by annarizzibramucci** [comment]
they are both just pathways, neither of them is a list
head(cutFs)
[1] "..../opt/anaconda3/bin/cutadapt"
> head(trimFs)
[1] "....../filtN/cutadapt/trimmed/cutadapt_R1_trim.fastq"



OK so the cutadapt code is searching in the wrong cutadapt folder for my sequences...
data.fp <- ("/...../AHG6G/filtN/cutadapt")
cutFs <- cutadapt[file.exists(cutadapt)]
cutRs <- cutadapt[file.exists(cutadapt)]
head(cutFs) 
"/.....opt/anaconda3/bin/cutadapt"

but even so it is just giving me a file pathway not a list of files...

**Comment by annarizzibramucci** [comment]
I actually think maybe I just need to rewrite the entire chunk of code here, it seems very confusing

**Comment by annarizzibramucci** [comment]
ok so I have rewritten my chunk of code and it now works if all of the files make it past the filtN step. So to input your step of filtN <- filtN[file.exists(filtN)], where/how do I input that so that my code won't stop if I am missing a sequence somewhere along the line (typically at the FiltN step, so I suppose I would put this code right after that step occurs?) Sorry for such a novice question...


data.fp <- ("....AHG6G/")
fnFs.filtN <- file.path(data.fp, "filtN", basename(fnFs)); # Put N-filterd files in filtN/ subdirectory
fnRs.filtN <- file.path(data.fp, "filtN", basename(fnRs));
filterAndTrim(fnFs, fnFs.filtN, fnRs, fnRs.filtN, maxN = 0, multithread = TRUE, compress=F);

data.fp <- ("...../AHG6G/filtN")
system2(cutadapt.file, args = "--version") # Run shell commands from R

path.cut <- file.path(data.fp, "cutadapt")
if(!dir.exists(path.cut)) dir.create(path.cut)
fnFs.cut <- file.path(path.cut, basename(fnFs))
fnRs.cut <- file.path(path.cut, basename(fnRs))

FWD.RC <- dada2:::rc(FWD)
REV.RC <- dada2:::rc(REV)
# Trim FWD and the reverse-complement of REV off of R1 (forward reads)
R1.flags <- paste("-g", FWD, "-a", REV.RC) 
# Trim REV and the reverse-complement of FWD off of R2 (reverse reads)
R2.flags <- paste("-G", REV, "-A", FWD.RC) 
# Run Cutadapt
for(i in seq_along(fnFs)) {
  system2(cutadapt, args = c(R1.flags, R2.flags, "-n", 2, # -n 2 required to remove FWD and REV from reads
                             "-o", fnFs.cut[i], "-p", fnRs.cut[i], # output files
                             fnFs.filtN[i], fnRs.filtN[i])) # input files
}

data.fp <- ("..../AHG6G/filtN/cutadapt")

primerHits <- function(primer, fn) {
    # Counts number of reads in which the primer is found
    nhits <- vcountPattern(primer, sread(readFastq(fn)), fixed = FALSE)
    return(sum(nhits > 0));
}
rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = fnFs.cut[[1]]), 
    FWD.ReverseReads = sapply(FWD.orients, primerHits, fn = fnRs.cut[[1]]), 
    REV.ForwardReads = sapply(REV.orients, primerHits, fn = fnFs.cut[[1]]), 
    REV.ReverseReads = sapply(REV.orients, primerHits, fn = fnRs.cut[[1]]));

data.fp <- ("...../AHG6G/filtN/cutadapt")
trimFs <- file.path(data.fp, "trimmed", paste0(sample.names, "_R1_trim.fastq"))
trimRs <- file.path(data.fp, "trimmed", paste0(sample.names, "_R2_trim.fastq"))
names(trimFs) <- sample.names
names(trimRs) <- sample.names

**Comment by benjjneb** [comment]
After running the `filterAndTrim` command that removed Ns, and leads to some samples being lost, you want to revise your lists of sample filenames to contain only those samples that passed that step, i.e. that have an existing output file. So, after that step...

```
passed.filtN <- file.exists(fnFs.filtN) # TRUE/FALSE vector of which samples passed the filter
fnFs.filtN <- fnFs.filtN[passed.filtN] # Keep only those samples that passed the filter
fnFs <- fnFs[passed.filtN] # Keep only those samples that passed the filter
```

**Comment by annarizzibramucci** [comment]
Thank you!
